File: ros2_ws/src/unitree_python_pkg/unitree_python_pkg/stand_and_sit_node.py
#!/usr/bin/env python3
"""Stand and Sit Node for Unitree Go1 real robot.

Sequence:
    1. INIT       - wait for the first /unitree_go1/low_state message
    2. STAND_UP   - linearly interpolate joints to standing position over
                    transition_steps control steps (unitree_guide semantics)
    3. HOLD       - hold standing position for hold_seconds seconds
    4. SIT_DOWN   - linearly interpolate joints to kennel position over
                    transition_steps steps
    5. PASSIVE    - damp-only command (Kp=0, Kd=3) then stop

Gains follow unitree_guide setRealStanceGain / State_Passive (real robot).
"""

import enum

import rclpy
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy

from unitree_legged_msgs.msg import BmsCmd, LowCmd, LowState, MotorCmd

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
PMSM_MODE = 0x0A  # servo mode used by the original controller

MOTOR_NAMES = [
    "FL_hip", "FL_thigh", "FL_calf",
    "FR_hip", "FR_thigh", "FR_calf",
    "RL_hip", "RL_thigh", "RL_calf",
    "RR_hip", "RR_thigh", "RR_calf",
]

# Standing target  (FL_hip, FL_thigh, FL_calf,  FR_hip, FR_thigh, FR_calf,
#                   RL_hip, RL_thigh, RL_calf,  RR_hip, RR_thigh, RR_calf)
# Source: State_FixedStand.h  _targetPos
STAND_POS = [
    0.0, 0.67, -1.3,
    0.0, 0.67, -1.3,
    0.0, 0.67, -1.3,
    0.0, 0.67, -1.3,
]

# Kennel (sitting) target
# Source: State_FixedKennel.h  _targetPos
KENNEL_POS = [
    0.35, 1.24, -2.81,
    -0.35, 1.24, -2.81,
    0.35, 1.24, -2.81,
    -0.35, 1.24, -2.81,
]

# Real-robot stance gains  (setRealStanceGain in LowlevelCmd.h)
# Per joint-type inside each leg: 0=hip, 1=thigh, 2=calf
REAL_STANCE_KP = [60.0, 40.0, 80.0]  # hip, thigh, calf
REAL_STANCE_KD = [5.0, 4.0, 7.0]  # hip, thigh, calf

# Passive gains  (State_Passive, real-robot branch)
PASSIVE_KP = 0.0
PASSIVE_KD = 3.0


# ---------------------------------------------------------------------------
# State machine
# ---------------------------------------------------------------------------
class Phase(enum.Enum):
    INIT = 0
    STAND_UP = 1
    HOLD = 2
    SIT_DOWN = 3
    PASSIVE = 4
    DONE = 5


class StandAndSitNode(Node):
    def __init__(self) -> None:
        super().__init__("stand_and_sit_node")

        # --- parameters ---
        self.declare_parameter("hold_seconds", 5.0)
        self.declare_parameter("transition_steps", 1000)
        self.declare_parameter("publish_rate", 500.0)
        self.declare_parameter("low_cmd_topic", "/unitree_go1/low_cmd")
        self.declare_parameter("low_state_topic", "/unitree_go1/low_state")
        self.declare_parameter("wait_for_low_state", True)

        self._hold_seconds = float(self._param("hold_seconds").double_value)
        self._transition_steps = int(
            self._param("transition_steps").integer_value
        )
        self._publish_rate = float(self._param("publish_rate").double_value)
        cmd_topic = self._param("low_cmd_topic").string_value
        state_topic = self._param("low_state_topic").string_value
        self._wait_for_low_state = self._param(
            "wait_for_low_state"
        ).bool_value

        if self._transition_steps < 1:
            self.get_logger().warning(
                "transition_steps must be >= 1. Forcing to 1."
            )
            self._transition_steps = 1
        if self._publish_rate <= 0.0:
            self.get_logger().warning(
                "publish_rate must be > 0. Forcing to 500.0 Hz."
            )
            self._publish_rate = 500.0
        if self._hold_seconds < 0.0:
            self.get_logger().warning(
                "hold_seconds must be >= 0. Forcing to 0.0s."
            )
            self._hold_seconds = 0.0

        self._dt = 1.0 / self._publish_rate

        # --- internal state ---
        self._phase = Phase.INIT
        self._step = 0  # step counter inside current transition
        self._hold_elapsed = 0.0  # seconds elapsed in HOLD phase
        self._passive_elapsed = 0.0
        self._start_pos = [0.0] * 12  # interpolation start snapshot
        self._current_joint_q = [0.0] * 12

        # --- publisher ---
        qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE,
            depth=1,
        )
        self._pub = self.create_publisher(LowCmd, cmd_topic, qos)

        # --- subscriber ---
        self._sub = self.create_subscription(
            LowState,
            state_topic,
            self._low_state_cb,
            10,
        )

        # --- pre-build the LowCmd skeleton ---
        self._low_cmd = self._make_low_cmd()

        self._log_startup_configuration(cmd_topic, state_topic)

        # --- if not waiting for low_state, start from KENNEL_POS ---
        if not self._wait_for_low_state:
            self.get_logger().info(
                "wait_for_low_state=False - starting stand-up from "
                "KENNEL_POS immediately."
            )
            self._start_pos = list(KENNEL_POS)
            self._apply_real_stance_gains()
            self._step = 0
            self._set_phase(Phase.STAND_UP, "starting from KENNEL_POS")

        # --- control timer ---
        self._timer = self.create_timer(
            self._dt, self._control_loop
        )

        if self._wait_for_low_state:
            self.get_logger().info(
                f"Waiting for first state on '{state_topic}' ..."
            )

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _param(self, name: str):
        return self.get_parameter(name).get_parameter_value()

    def _log_startup_configuration(
        self,
        cmd_topic: str,
        state_topic: str,
    ) -> None:
        self.get_logger().info(
            "\n".join(
                [
                    "Stand-and-sit configuration:",
                    f"  publish LowCmd topic: {cmd_topic}",
                    f"  subscribe LowState topic: {state_topic}",
                    f"  hold_seconds: {self._hold_seconds:.3f}s",
                    f"  transition_steps: {self._transition_steps}",
                    (
                        "  transition_duration: "
                        f"{self._transition_steps * self._dt:.3f}s"
                    ),
                    f"  publish_rate: {self._publish_rate:.3f} Hz",
                    "  passive_duration: 1.000s",
                    f"  wait_for_low_state: {self._wait_for_low_state}",
                    self._format_joint_targets("STAND_POS", STAND_POS),
                    self._format_joint_targets("KENNEL_POS", KENNEL_POS),
                    self._format_stance_gains(),
                ]
            )
        )

    def _format_joint_targets(self, label: str, values: list) -> str:
        lines = [f"  {label}:"]
        for leg_name, offset in (
            ("FL", 0),
            ("FR", 3),
            ("RL", 6),
            ("RR", 9),
        ):
            lines.append(
                "    "
                f"{leg_name}: "
                f"hip={values[offset]: .3f}  "
                f"thigh={values[offset + 1]: .3f}  "
                f"calf={values[offset + 2]: .3f}"
            )
        return "\n".join(lines)

    def _format_stance_gains(self) -> str:
        return (
            "  REAL_STANCE gains: "
            f"hip(kp={REAL_STANCE_KP[0]:.1f}, kd={REAL_STANCE_KD[0]:.1f})  "
            f"thigh(kp={REAL_STANCE_KP[1]:.1f}, kd={REAL_STANCE_KD[1]:.1f})  "
            f"calf(kp={REAL_STANCE_KP[2]:.1f}, kd={REAL_STANCE_KD[2]:.1f})  "
            f"PASSIVE(kp={PASSIVE_KP:.1f}, kd={PASSIVE_KD:.1f})"
        )

    def _format_float_array(self, values: list) -> str:
        return "[" + ", ".join(f"{float(value):.3f}" for value in values) + "]"

    def _log_low_state(self, msg: LowState) -> None:
        lines = [
            "First LowState data:",
            (
                "  "
                f"head={list(msg.head)}  "
                f"level_flag={msg.level_flag}  "
                f"band_width={msg.band_width}  "
                f"tick={msg.tick}  "
                f"crc={msg.crc}"
            ),
            f"  foot_force={self._format_float_array(msg.foot_force)}",
            f"  foot_force_est={self._format_float_array(msg.foot_force_est)}",
            "  idx  motor       q(rad)    dq(rad/s)  tau_est(Nm)  temp(C)",
        ]
        for i, motor in enumerate(msg.motor_state):
            lines.append(
                "  "
                f"{i:>2}   {MOTOR_NAMES[i]:<9}  "
                f"{motor.q:>8.3f}  "
                f"{motor.dq:>9.3f}  "
                f"{motor.tau_est:>11.3f}  "
                f"{motor.temperature:>7}"
            )
        self.get_logger().info("\n".join(lines))

    def _log_command(self, title: str) -> None:
        lines = [
            title,
            "  idx  motor       q(rad)       kp       kd",
        ]
        for i, motor in enumerate(self._low_cmd.motor_cmd):
            lines.append(
                "  "
                f"{i:>2}   {MOTOR_NAMES[i]:<9}  "
                f"{motor.q:>8.3f}  "
                f"{motor.kp:>7.1f}  "
                f"{motor.kd:>7.1f}"
            )
        self.get_logger().info("\n".join(lines))

    def _set_phase(self, phase: Phase, reason: str) -> None:
        self._phase = phase
        self.get_logger().info(f"Phase -> {phase.name}: {reason}")

    def _make_low_cmd(self) -> LowCmd:
        cmd = LowCmd()
        cmd.head = [0xFE, 0xEF]
        cmd.level_flag = 0x01
        cmd.frame_reserve = 0
        cmd.sn = [0, 0]
        cmd.version = [1, 0]
        cmd.band_width = int(round(self._publish_rate))
        cmd.motor_cmd = [MotorCmd() for _ in range(12)]
        cmd.bms = BmsCmd()
        cmd.bms.off = 0
        cmd.bms.reserve = [0, 0, 0]
        cmd.crc = 0
        for i in range(12):
            cmd.motor_cmd[i].mode = PMSM_MODE
            cmd.motor_cmd[i].q = 0.0
            cmd.motor_cmd[i].dq = 0.0
            cmd.motor_cmd[i].tau = 0.0
            cmd.motor_cmd[i].kp = 0.0
            cmd.motor_cmd[i].kd = 0.0
        return cmd

    def _apply_real_stance_gains(self) -> None:
        """Set Kp/Kd for each motor matching setRealStanceGain."""
        for leg in range(4):
            for jt in range(3):
                idx = leg * 3 + jt
                self._low_cmd.motor_cmd[idx].mode = PMSM_MODE
                self._low_cmd.motor_cmd[idx].kp = REAL_STANCE_KP[jt]
                self._low_cmd.motor_cmd[idx].kd = REAL_STANCE_KD[jt]
                self._low_cmd.motor_cmd[idx].dq = 0.0
                self._low_cmd.motor_cmd[idx].tau = 0.0

    def _apply_passive_gains(self) -> None:
        """Set damping-only gains matching State_Passive (real robot)."""
        for i in range(12):
            self._low_cmd.motor_cmd[i].mode = PMSM_MODE
            self._low_cmd.motor_cmd[i].kp = PASSIVE_KP
            self._low_cmd.motor_cmd[i].kd = PASSIVE_KD
            self._low_cmd.motor_cmd[i].q = 0.0
            self._low_cmd.motor_cmd[i].dq = 0.0
            self._low_cmd.motor_cmd[i].tau = 0.0

    def _lerp_and_publish(self, start: list, target: list) -> None:
        """Publish an interpolated command between start and target."""
        alpha = min(float(self._step) / float(self._transition_steps), 1.0)
        for i in range(12):
            self._low_cmd.motor_cmd[i].q = (
                (1.0 - alpha) * start[i] + alpha * target[i]
            )
        self._pub.publish(self._low_cmd)

    def _publish_position_command(self, target: list) -> None:
        for i in range(12):
            self._low_cmd.motor_cmd[i].q = target[i]
        self._pub.publish(self._low_cmd)

    # ------------------------------------------------------------------
    # Subscriber callback
    # ------------------------------------------------------------------

    def _low_state_cb(self, msg: LowState) -> None:
        for i in range(12):
            self._current_joint_q[i] = float(msg.motor_state[i].q)

        if self._phase == Phase.INIT:
            self.get_logger().info("Got first low_state - starting stand-up.")
            self._log_low_state(msg)
            # Snapshot start positions and configure gains
            self._start_pos = list(self._current_joint_q)
            self._apply_real_stance_gains()
            self._step = 0
            self._set_phase(Phase.STAND_UP, "first LowState received")

    # ------------------------------------------------------------------
    # Control loop (runs at publish_rate Hz)
    # ------------------------------------------------------------------

    def _control_loop(self) -> None:
        if self._phase == Phase.INIT:
            return  # waiting for first state

        if self._phase == Phase.STAND_UP:
            self._step += 1
            self._lerp_and_publish(self._start_pos, STAND_POS)
            if self._step >= self._transition_steps:
                self._publish_position_command(STAND_POS)
                self._log_command("Published STAND_UP target command:")
                self.get_logger().info(
                    "Standing up complete. "
                    f"Holding for {self._hold_seconds}s ..."
                )
                self._hold_elapsed = 0.0
                self._set_phase(Phase.HOLD, "stand-up complete")

        elif self._phase == Phase.HOLD:
            # Keep publishing the stand position
            for i in range(12):
                self._low_cmd.motor_cmd[i].q = STAND_POS[i]
            self._pub.publish(self._low_cmd)

            self._hold_elapsed += self._dt
            if self._hold_elapsed >= self._hold_seconds:
                self.get_logger().info("Hold complete. Sitting down ...")
                # Start from currently commanded pose for continuity.
                self._start_pos = [
                    float(motor.q) for motor in self._low_cmd.motor_cmd
                ]
                self._step = 0
                self._set_phase(Phase.SIT_DOWN, "hold complete")

        elif self._phase == Phase.SIT_DOWN:
            self._step += 1
            self._lerp_and_publish(self._start_pos, KENNEL_POS)
            if self._step >= self._transition_steps:
                self._publish_position_command(KENNEL_POS)
                self._log_command("Published SIT_DOWN target command:")
                self.get_logger().info(
                    "Sit-down complete. Switching to passive mode."
                )
                self._apply_passive_gains()
                self._passive_elapsed = 0.0
                self._pub.publish(self._low_cmd)
                self._log_command("Published PASSIVE command:")
                self._set_phase(Phase.PASSIVE, "sit-down complete")

        elif self._phase == Phase.PASSIVE:
            self._pub.publish(self._low_cmd)
            self._passive_elapsed += self._dt
            if self._passive_elapsed >= 1.0:
                self._log_command("Final PASSIVE command:")
                self.get_logger().info("Passive phase done. Shutting down.")
                self._timer.cancel()
                self._set_phase(Phase.DONE, "passive phase complete")
                rclpy.shutdown()

        # Phase.DONE: timer already cancelled, nothing to do


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main(args=None) -> None:
    rclpy.init(args=args)
    try:
        node = StandAndSitNode()
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_ws/src/unitree_cpp_pkg/src/stand_and_sit_node.cpp
#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <memory>
#include <sstream>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/qos.hpp"
#include "unitree_legged_msgs/msg/bms_cmd.hpp"
#include "unitree_legged_msgs/msg/low_cmd.hpp"
#include "unitree_legged_msgs/msg/low_state.hpp"

namespace {

constexpr uint8_t kPmsmMode = 0x0A;
constexpr double kPassiveDurationSeconds = 1.0;

constexpr std::array<double, 12> kStandPos = {
    0.0, 0.67, -1.3,
    0.0, 0.67, -1.3,
    0.0, 0.67, -1.3,
    0.0, 0.67, -1.3,
};

constexpr std::array<double, 12> kKennelPos = {
    0.35, 1.24, -2.81,
    -0.35, 1.24, -2.81,
    0.35, 1.24, -2.81,
    -0.35, 1.24, -2.81,
};

constexpr std::array<double, 3> kRealStanceKp = {60.0, 40.0, 80.0};
constexpr std::array<double, 3> kRealStanceKd = {5.0, 4.0, 7.0};
constexpr double kPassiveKp = 0.0;
constexpr double kPassiveKd = 3.0;

enum class Phase {
  INIT = 0,
  STAND_UP = 1,
  HOLD = 2,
  SIT_DOWN = 3,
  PASSIVE = 4,
  DONE = 5,
};

class StandAndSitNode : public rclcpp::Node {
 public:
  StandAndSitNode()
  : Node("stand_and_sit_node"),
    phase_(Phase::INIT),
    step_(0),
    hold_elapsed_(0.0),
    passive_elapsed_(0.0) {
    this->declare_parameter<double>("hold_seconds", 5.0);
    this->declare_parameter<int>("transition_steps", 1000);
    this->declare_parameter<double>("publish_rate", 500.0);
    this->declare_parameter<int>("init_state_samples", 20);
    this->declare_parameter<std::string>("low_cmd_topic", "/unitree_go1/low_cmd");
    this->declare_parameter<std::string>("low_state_topic", "/unitree_go1/low_state");
    this->declare_parameter<bool>("wait_for_low_state", true);

    hold_seconds_ = this->get_parameter("hold_seconds").as_double();
    transition_steps_ = this->get_parameter("transition_steps").as_int();
    publish_rate_ = this->get_parameter("publish_rate").as_double();
    init_state_samples_ = this->get_parameter("init_state_samples").as_int();
    low_cmd_topic_ = this->get_parameter("low_cmd_topic").as_string();
    low_state_topic_ = this->get_parameter("low_state_topic").as_string();
    wait_for_low_state_ = this->get_parameter("wait_for_low_state").as_bool();

    if (transition_steps_ < 1) {
      RCLCPP_WARN(this->get_logger(), "transition_steps must be >= 1. Forcing to 1.");
      transition_steps_ = 1;
    }
    if (publish_rate_ <= 0.0) {
      RCLCPP_WARN(this->get_logger(), "publish_rate must be > 0. Forcing to 500.0 Hz.");
      publish_rate_ = 500.0;
    }
    if (hold_seconds_ < 0.0) {
      RCLCPP_WARN(this->get_logger(), "hold_seconds must be >= 0. Forcing to 0.0s.");
      hold_seconds_ = 0.0;
    }
    if (init_state_samples_ < 1) {
      RCLCPP_WARN(this->get_logger(), "init_state_samples must be >= 1. Forcing to 1.");
      init_state_samples_ = 1;
    }

    dt_ = 1.0 / publish_rate_;
    start_pos_.fill(0.0);
    current_joint_q_.fill(0.0);
    init_q_sum_.fill(0.0);

    rclcpp::QoS qos(1);
    qos.reliable();
    qos.durability_volatile();
    pub_ = this->create_publisher<unitree_legged_msgs::msg::LowCmd>(low_cmd_topic_, qos);
    sub_ = this->create_subscription<unitree_legged_msgs::msg::LowState>(
      low_state_topic_,
      10,
      std::bind(&StandAndSitNode::low_state_cb, this, std::placeholders::_1));

    make_low_cmd();
    log_configuration();

    if (!wait_for_low_state_) {
      RCLCPP_INFO(
        this->get_logger(),
        "wait_for_low_state=false: starting stand-up from KENNEL_POS immediately.");
      start_pos_ = kKennelPos;
      apply_real_stance_gains();
      step_ = 0;
      set_phase(Phase::STAND_UP, "starting from KENNEL_POS");
    }

    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(dt_),
      std::bind(&StandAndSitNode::control_loop, this));

    if (wait_for_low_state_) {
      RCLCPP_INFO(
        this->get_logger(),
        "Waiting for first state on '%s'...",
        low_state_topic_.c_str());
    }
  }

 private:
  void log_configuration() {
    std::ostringstream oss;
    oss << "Stand-and-sit configuration:\n"
        << "  publish LowCmd topic: " << low_cmd_topic_ << "\n"
        << "  subscribe LowState topic: " << low_state_topic_ << "\n"
        << "  hold_seconds: " << hold_seconds_ << "s\n"
        << "  transition_steps: " << transition_steps_ << "\n"
        << "  transition_duration: " << (static_cast<double>(transition_steps_) * dt_) << "s\n"
        << "  init_state_samples: " << init_state_samples_ << "\n"
        << "  publish_rate: " << publish_rate_ << " Hz\n"
        << "  passive_duration: " << kPassiveDurationSeconds << "s\n"
        << "  wait_for_low_state: " << (wait_for_low_state_ ? "true" : "false");
    RCLCPP_INFO(this->get_logger(), "%s", oss.str().c_str());
  }

  void set_phase(Phase new_phase, const std::string & reason) {
    phase_ = new_phase;
    RCLCPP_INFO(this->get_logger(), "Phase -> %d: %s", static_cast<int>(phase_), reason.c_str());
  }

  void make_low_cmd() {
    low_cmd_ = unitree_legged_msgs::msg::LowCmd();
    low_cmd_.head[0] = 0xFE;
    low_cmd_.head[1] = 0xEF;
    low_cmd_.level_flag = 0x01;
    low_cmd_.frame_reserve = 0;
    low_cmd_.sn[0] = 0;
    low_cmd_.sn[1] = 0;
    low_cmd_.version[0] = 1;
    low_cmd_.version[1] = 0;
    low_cmd_.band_width = static_cast<uint16_t>(std::lround(publish_rate_));
    low_cmd_.bms = unitree_legged_msgs::msg::BmsCmd();
    low_cmd_.bms.off = 0;
    low_cmd_.bms.reserve = {0, 0, 0};
    low_cmd_.reserve = 0;
    low_cmd_.crc = 0;

    for (std::size_t i = 0; i < low_cmd_.motor_cmd.size(); ++i) {
      low_cmd_.motor_cmd[i].mode = kPmsmMode;
      low_cmd_.motor_cmd[i].q = 0.0F;
      low_cmd_.motor_cmd[i].dq = 0.0F;
      low_cmd_.motor_cmd[i].tau = 0.0F;
      low_cmd_.motor_cmd[i].kp = 0.0F;
      low_cmd_.motor_cmd[i].kd = 0.0F;
      low_cmd_.motor_cmd[i].reserve = {0U, 0U, 0U};
    }
  }

  void apply_real_stance_gains() {
    for (int leg = 0; leg < 4; ++leg) {
      for (int jt = 0; jt < 3; ++jt) {
        const int idx = leg * 3 + jt;
        low_cmd_.motor_cmd[idx].mode = kPmsmMode;
        low_cmd_.motor_cmd[idx].kp = static_cast<float>(kRealStanceKp[jt]);
        low_cmd_.motor_cmd[idx].kd = static_cast<float>(kRealStanceKd[jt]);
        low_cmd_.motor_cmd[idx].dq = 0.0F;
        low_cmd_.motor_cmd[idx].tau = 0.0F;
      }
    }
  }

  void apply_passive_gains() {
    for (int i = 0; i < 12; ++i) {
      low_cmd_.motor_cmd[i].mode = kPmsmMode;
      low_cmd_.motor_cmd[i].kp = static_cast<float>(kPassiveKp);
      low_cmd_.motor_cmd[i].kd = static_cast<float>(kPassiveKd);
      low_cmd_.motor_cmd[i].q = 0.0F;
      low_cmd_.motor_cmd[i].dq = 0.0F;
      low_cmd_.motor_cmd[i].tau = 0.0F;
    }
  }

  void apply_init_safe_command() {
    for (int i = 0; i < 12; ++i) {
      low_cmd_.motor_cmd[i].mode = kPmsmMode;
      low_cmd_.motor_cmd[i].q = 0.0F;
      low_cmd_.motor_cmd[i].dq = 0.0F;
      low_cmd_.motor_cmd[i].kp = 0.0F;
      low_cmd_.motor_cmd[i].kd = 0.0F;
      low_cmd_.motor_cmd[i].tau = 0.0F;
    }
  }

  void lerp_and_publish(const std::array<double, 12> & start, const std::array<double, 12> & target) {
    const double alpha = std::min(static_cast<double>(step_) / static_cast<double>(transition_steps_), 1.0);
    for (int i = 0; i < 12; ++i) {
      const double q = (1.0 - alpha) * start[i] + alpha * target[i];
      low_cmd_.motor_cmd[i].q = static_cast<float>(q);
    }
    pub_->publish(low_cmd_);
  }

  void publish_position_command(const std::array<double, 12> & target) {
    for (int i = 0; i < 12; ++i) {
      low_cmd_.motor_cmd[i].q = static_cast<float>(target[i]);
    }
    pub_->publish(low_cmd_);
  }

  bool has_valid_sensor_state() {
    if (!has_low_state_) {
      return false;
    }

    if (latest_low_state_.tick == 0U) {
      RCLCPP_WARN_THROTTLE(
        this->get_logger(),
        *this->get_clock(),
        2000,
        "LowState tick is zero; using joint values and skipping duplicate-tick filtering.");
    }

    double max_abs_q = 0.0;
    for (int i = 0; i < 12; ++i) {
      const double q = static_cast<double>(latest_low_state_.motor_state[i].q);
      if (!std::isfinite(q) || std::abs(q) > 10.0) {
        return false;
      }
      max_abs_q = std::max(max_abs_q, std::abs(q));
    }
    return max_abs_q > 0.05;
  }

  void low_state_cb(const unitree_legged_msgs::msg::LowState::SharedPtr msg) {
    latest_low_state_ = *msg;
    has_low_state_ = true;

    for (int i = 0; i < 12; ++i) {
      current_joint_q_[i] = static_cast<double>(latest_low_state_.motor_state[i].q);
    }
  }

  void control_loop() {
    if (phase_ == Phase::DONE) {
      return;
    }

    if (phase_ == Phase::INIT) {
      if (wait_for_low_state_ && !has_low_state_) {
        apply_init_safe_command();
        pub_->publish(low_cmd_);
        RCLCPP_WARN_THROTTLE(
          this->get_logger(),
          *this->get_clock(),
          2000,
          "Waiting for first LowState while keeping INIT command stream alive.");
        return;
      }

      if (!has_valid_sensor_state()) {
        apply_init_safe_command();
        pub_->publish(low_cmd_);
        RCLCPP_WARN_THROTTLE(
          this->get_logger(),
          *this->get_clock(),
          2000,
          "Ignoring invalid LowState while initializing start posture.");
        return;
      }

      if (latest_low_state_.tick != 0U) {
        if (latest_low_state_.tick == last_tick_) {
          apply_init_safe_command();
          pub_->publish(low_cmd_);
          return;
        }
        last_tick_ = latest_low_state_.tick;
      }

      for (int i = 0; i < 12; ++i) {
        init_q_sum_[i] += current_joint_q_[i];
      }
      ++init_sample_count_;

      if (init_sample_count_ < init_state_samples_) {
        apply_init_safe_command();
        pub_->publish(low_cmd_);
        return;
      }

      for (int i = 0; i < 12; ++i) {
        start_pos_[i] = init_q_sum_[i] / static_cast<double>(init_sample_count_);
      }

      apply_real_stance_gains();
      publish_position_command(start_pos_);
      step_ = 0;
      set_phase(Phase::STAND_UP, "sensor initialization complete");
      return;
    }

    if (phase_ == Phase::STAND_UP) {
      ++step_;
      lerp_and_publish(start_pos_, kStandPos);
      if (step_ >= transition_steps_) {
        publish_position_command(kStandPos);
        hold_elapsed_ = 0.0;
        set_phase(Phase::HOLD, "stand-up complete");
      }
      return;
    }

    if (phase_ == Phase::HOLD) {
      publish_position_command(kStandPos);
      hold_elapsed_ += dt_;
      if (hold_elapsed_ >= hold_seconds_) {
        for (int i = 0; i < 12; ++i) {
          start_pos_[i] = current_joint_q_[i];
        }
        step_ = 0;
        set_phase(Phase::SIT_DOWN, "hold complete");
      }
      return;
    }

    if (phase_ == Phase::SIT_DOWN) {
      ++step_;
      lerp_and_publish(start_pos_, kKennelPos);
      if (step_ >= transition_steps_) {
        publish_position_command(kKennelPos);
        apply_passive_gains();
        passive_elapsed_ = 0.0;
        pub_->publish(low_cmd_);
        set_phase(Phase::PASSIVE, "sit-down complete");
      }
      return;
    }

    if (phase_ == Phase::PASSIVE) {
      pub_->publish(low_cmd_);
      passive_elapsed_ += dt_;
      if (passive_elapsed_ >= kPassiveDurationSeconds) {
        timer_->cancel();
        set_phase(Phase::DONE, "passive phase complete");
        rclcpp::shutdown();
      }
    }
  }

  double hold_seconds_{5.0};
  int transition_steps_{1000};
  double publish_rate_{500.0};
  int init_state_samples_{20};
  double dt_{0.002};
  std::string low_cmd_topic_;
  std::string low_state_topic_;
  bool wait_for_low_state_{true};

  Phase phase_;
  int step_;
  double hold_elapsed_;
  double passive_elapsed_;
  int init_sample_count_{0};
  uint32_t last_tick_{0};
  bool has_low_state_{false};
  std::array<double, 12> start_pos_;
  std::array<double, 12> current_joint_q_;
  std::array<double, 12> init_q_sum_;
  unitree_legged_msgs::msg::LowState latest_low_state_;

  unitree_legged_msgs::msg::LowCmd low_cmd_;

  rclcpp::Publisher<unitree_legged_msgs::msg::LowCmd>::SharedPtr pub_;
  rclcpp::Subscription<unitree_legged_msgs::msg::LowState>::SharedPtr sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

}  // namespace

int main(int argc, char ** argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<StandAndSitNode>();
  rclcpp::spin(node);
  if (rclcpp::ok()) {
    rclcpp::shutdown();
  }
  return 0;
}
